Source/WebCore/Modules/mediastream/VideoTrackGenerator.cpp

```diff
@@ -27,24 +27,166 @@
 
 #if ENABLE(MEDIA_STREAM)
 
+#include "Exception.h"
+#include "JSWebCodecsVideoFrame.h"
 #include "MediaStreamTrack.h"
 #include "WritableStream.h"
+#include "WritableStreamSink.h"
 
 namespace WebCore {
 
 WTF_MAKE_ISO_ALLOCATED_IMPL(VideoTrackGenerator);
 
-VideoTrackGenerator::VideoTrackGenerator(ScriptExecutionContext&)
+ExceptionOr<Ref<VideoTrackGenerator>> VideoTrackGenerator::create(ScriptExecutionContext& context)
 {
+    auto source = Source::create();
+    auto sink = Sink::create(Ref { source });
+    auto writableOrException = WritableStream::create(*JSC::jsCast<JSDOMGlobalObject*>(context.globalObject()), Ref { sink });
+
+    if (writableOrException.hasException())
+        return writableOrException.releaseException();
+
+    // FIXME: Maybe we should have the writable buffer frames until the source is actually started.
+    callOnMainThread([source] {
+        source->start();
+    });
+
+    auto track = MediaStreamTrack::create(context, MediaStreamTrackPrivate::create(Logger::create(&context), WTFMove(source), [identifier = context.identifier()](Function<void()>&& task) {
+        ScriptExecutionContext::postTaskTo(identifier, [task = WTFMove(task)] (auto&) mutable {
+            task();
+        });
+    }));
+
+    return adoptRef(*new VideoTrackGenerator(WTFMove(sink), writableOrException.releaseReturnValue(), WTFMove(track)));
+}
+
+VideoTrackGenerator::VideoTrackGenerator(Ref<Sink>&& sink, Ref<WritableStream>&& writable, Ref<MediaStreamTrack>&& track)
+    : m_sink(WTFMove(sink))
+    , m_writable(WTFMove(writable))
+    , m_track(WTFMove(track))
+{
+    RealtimeMediaSourceSupportedConstraints supportedConstraints;
+    supportedConstraints.setSupportsWidth(true);
+    supportedConstraints.setSupportsHeight(true);
+
+    RealtimeMediaSourceSettings settings;
+    settings.setSupportedConstraints(supportedConstraints);
+    settings.setWidth(0);
+    settings.setHeight(0);
+
+    RealtimeMediaSourceCapabilities capabilities { supportedConstraints };
+    capabilities.setWidth({ 0, 0 });
+    capabilities.setHeight({ 0, 0 });
+
+    m_track->privateTrack().initializeSettings(WTFMove(settings));
+    m_track->privateTrack().initializeCapabilities(WTFMove(capabilities));
```

Comment:
It would be nice to remove `track->privateTrack()` eventually. Can you put these methods on MediaStreamTrack and have it call the private track?

Reply:
Will update the code.